oracle/src/events/processAffirmationRequests/estimateGas.js

```diff
@@ -4,19 +4,30 @@ const logger = require('../../services/logger').child({
   module: 'processAffirmationRequests:estimateGas'
 })
 
-async function estimateGas({ web3, homeBridge, validatorContract, recipient, value, txHash, address }) {
+async function estimateGas({ web3, homeBridge, validatorContract, recipient, value, nonce, address }) {
   try {
-    return await homeBridge.methods.executeAffirmation(recipient, value, txHash).estimateGas({
+    return await homeBridge.methods.executeAffirmation(recipient, value, nonce).estimateGas({
       from: address
     })
   } catch (e) {
     if (e instanceof HttpListProviderError) {
       throw e
     }
 
-    const messageHash = web3.utils.soliditySha3(recipient, value, txHash)
+    const messageHash = web3.utils.soliditySha3(recipient, value, nonce)
     const senderHash = web3.utils.soliditySha3(address, messageHash)
 
+    const isHashiMandatory = await homeBridge.methods.HASHI_IS_MANDATORY().call()
+    const isHashiEnabled = await homeBridge.methods.HASHI_IS_ENABLED().call()
+
+    if (isHashiMandatory === true && isHashiEnabled === true) {
+      // Check if msg is approved by Hashi
+      const isApprovedByHashi = await homeBridge.methods.isApprovedByHashi(messageHash)
+      if (!isApprovedByHashi) {
+        throw new NotApprovedByHashiError(e.message)
```

Comment:
I don't know if it does make sense to throw an error here. IMHO you should just return or wait for the hashi approval.

Reply:
The validator will wait for `requiredBlockConfirmations` blocks before processing the tx. It is the maximum block number required for the Hashi oracle to store hash. Ideally, after `requiredBlockConfirmations` blocks, the Hashi threshold should be reached.

Reply:
i don't think so. You don't know when the quorum for a message will be reached

Reply:
Retry logic added in [6852ab3](https://github.com/gnosischain/tokenbridge/pull/5/commits/6852ab37504fdc1eb646cbc0827a522975b01d7a)